### PR TITLE
feat: implement \R meta-command for prompt change

### LINKED ISSUE #346: Implement \R meta-command for prompt change
## Overview

Implement the `\R` meta-command to change the prompt string, compatible with Google Cloud Spanner CLI.

## Specification

- Syntax: `\R <prompt_string>`
- Changes the prompt to the specified string
- Should persist for the session

## Implementation notes

- spanner-mycli already has sophisticated prompt support with `%` expansion
- Consider how `\R` interacts with existing prompt system
- Should update the `CLI_PROMPT` system variable
- May need to support both simple strings and expansion patterns

## Priority

Low priority - spanner-mycli already has `--prompt` flag and advanced prompt features

## References

- Parent issue: #341
- [Google Cloud Spanner CLI documentation](https://cloud.google.com/spanner/docs/spanner-cli#supported-meta-commands)

### PR DESCRIPTION
**Pull request by apstndb**
## Summary
- Implement `\R <prompt_string>` meta-command to change the prompt dynamically
- Compatible with Google Cloud Spanner CLI behavior
- Updates the `CLI_PROMPT` system variable with automatic trailing space

## Changes
- Add `PromptMetaCommand` struct that implements `MetaCommandStatement` interface
- Update `ParseMetaCommand` to handle `\R` command syntax
- Add comprehensive unit tests for parsing and execution
- Add integration tests for interactive mode behavior
- Update documentation in README.md and docs/meta_commands.md

### Key Design Decisions

1. **Trailing Space Handling**: The `\R` command automatically adds a trailing space to the prompt string for better UX and compatibility with Google Cloud Spanner CLI. This ensures proper separation between the prompt and user input.

2. **Interface Design**: All meta-commands implement `MetaCommandStatement` which embeds `Statement`, ensuring type safety and consistency. This allows us to verify at compile time that all meta-commands are valid statements.

3. **Execution Flow**: The `\R` command updates the prompt by calling `systemVariables.Set("CLI_PROMPT", promptWithSpace)`, maintaining consistency with the `SET CLI_PROMPT = ...` statement.

4. **Error Handling**: Empty prompt commands (`\R` without arguments) return a clear error message, matching the pattern of other meta-commands.

5. **Input Processing**: The input string is trimmed (consistent with other meta-commands), but a space is always appended when setting the prompt. This means `\R spanner>` results in `CLI_PROMPT="spanner> "`.

## Testing
- Unit tests cover all parsing scenarios (simple prompt, percent expansion, empty prompt error)
- Integration tests verify interactive mode behavior
- All existing tests pass with `make check`

## Manual Test Instructions
1. Start spanner-mycli: `./spanner-mycli --embedded-emulator -p test -i test -d test`
2. Test basic prompt change: `\R mycli>`
3. Test with percent expansion: `\R [%p/%i/%d]>`
4. Verify change: `SHOW VARIABLE CLI_PROMPT;` (should show trailing space)
5. Test error case: `\R` (should show error)

Fixes #346

### CODE REVIEW

File: meta_commands.go [hunk 1]
```diff
@@ -114,6 +113,11 @@ func ParseMetaCommand(input string) (Statement, error) {
 			return nil, errors.New("\\. requires exactly one filename")
 		}
 		return &SourceMetaCommand{FilePath: words[0]}, nil
+	case "R":
+		if args == "" {
+			return nil, errors.New("\\R requires a prompt string")
+		}
+		return &PromptMetaCommand{PromptString: args}, nil
```
[the remark this answers is not in the document]
Comment (apstndb) [reply meta_commands.go]: Thank you for the feedback!

Fixed in commit a8589f3.

File: meta_commands.go [hunk 2]
```diff
@@ -114,6 +113,12 @@ func ParseMetaCommand(input string) (Statement, error) {
 			return nil, errors.New("\\. requires exactly one filename")
 		}
 		return &SourceMetaCommand{FilePath: words[0]}, nil
+	case "R":
+		trimmedArgs := strings.TrimSpace(args)
+		if trimmedArgs == "" {
+			return nil, errors.New("\\R requires a prompt string")
+		}
+		return &PromptMetaCommand{PromptString: trimmedArgs}, nil
```
[the remark this answers is not in the document]
Comment (apstndb) [reply meta_commands.go:121]: Thank you for the feedback!

Fixed in commit ff1f00f.

File: integration_meta_command_test.go [hunk 3]
```diff
@@ -249,9 +263,130 @@ SELECT "foo" AS s;`
...
+			if _, ok := err.(*ExitCodeError); !ok {
+				t.Errorf("RunInteractive() returned an unexpected error = %v", err)
+			}
+		}
+
+		// Verify the prompt was changed (expansion happens during display, with trailing space added)
+		if sysVars.Prompt != "[%p/%i/%d]> " {
+			t.Errorf("Expected prompt to be '[%%p/%%i/%%d]> ', got %q", sysVars.Prompt)
+		}
+	})
```
[the remark this answers is not in the document]
Comment (apstndb) [reply integration_meta_command_test.go]: Thank you for the feedback!

Fixed in commit 04c2a42.

**Comment by apstndb** [comment]
## Implementation Notes for Future Reference

This PR successfully implements the `\R` meta-command with some important design decisions:

### 1. Automatic Trailing Space
The implementation automatically adds a trailing space to all prompts set via `\R`. This deviates from the literal specification but ensures compatibility with Google Cloud Spanner CLI behavior, where prompts naturally have a space between the prompt and user input.

### 2. Security Fix Included
During implementation, we discovered and fixed a critical security vulnerability where invalid percent sequences (like `%z`) would cause a panic in `getInterpolatedPrompt`. This is now handled gracefully with comprehensive test coverage.

### 3. Test Coverage Considerations
- **Unit tests**: Comprehensive coverage including edge cases and invalid percent sequences
- **Integration tests**: Limited by the lack of a real Spanner connection. The tests verify that prompts are stored correctly but cannot verify actual percent expansion (e.g., `%p`, `%i`, `%d`) without project/instance/database context.

### 4. Development Artifacts
Temporary test files were created during development (in `tmp/`) but were intentionally not committed as they were only needed for debugging during implementation.

### Implementation Highlights
- Clean integration with existing prompt system and system variables
- Follows established patterns for meta-commands
- Maintains backward compatibility with existing prompt features
- All review feedback has been addressed

The implementation is complete and ready for merge. 🎉
EOF < /dev/null
